File: Days 31 - 40/Day 39/ConwaysGameOfLife.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading;

namespace DailyCodingProblem
{
	internal enum CellStates
	{
		Alive = '*',
		Dead = '.'
	}

	internal struct Cell
	{
		public static int minX = int.MaxValue;
		public static int maxX = int.MinValue;
		public static int minY = int.MaxValue;
		public static int maxY = int.MinValue;

		public int x;
		public int y;

		public Cell(int x, int y)
		{
			this.x = x;
			this.y = y;
		}

		public static bool operator ==(Cell lhs, Cell rhs) => lhs.x == rhs.x && lhs.y == rhs.y;
		public static bool operator !=(Cell lhs, Cell rhs) => !(lhs == rhs);

		public static bool operator <(Cell lhs, Cell rhs) => lhs.x != rhs.x ? lhs.x < rhs.x : lhs.y < rhs.y;
		public static bool operator >(Cell lhs, Cell rhs) => lhs.x != rhs.x ? lhs.x > rhs.x : lhs.y > rhs.y;

		public override bool Equals(object obj)
		{
			if (!(obj is Cell))
			{
				return false;
			}

			Cell cell = (Cell)obj;
			return x == cell.x && y == cell.y;
		}

		public override int GetHashCode() => base.GetHashCode();
	}

	internal class GameOfLife
	{
		private List<Cell> cells;
		private List<StringBuilder> board = new List<StringBuilder>();

		public GameOfLife(List<Cell> cells)
		{
			this.cells = cells;
		}

		public void Play(int iterations = 100, int millisecondDelta = 500)
		{
			UpdateCellMinAndMax();
			CellsToBoard();

			PrintBoard();
			Thread.Sleep(millisecondDelta);

			for (int i = 0; i < iterations; i++)
			{
				board = IterateCells();
				PrintBoard(i + 1);

				Thread.Sleep(millisecondDelta);
			}
		}

		private List<StringBuilder> IterateCells()
		{
			UpdateCellMinAndMax();

			List<Cell> newCells = new List<Cell>();
			List<Cell> deadCells = new List<Cell>();

			if (cells.Count > 0)
			{
				CellsToBoard();

				for (int i = 0; i < board.Count; i++)
				{
					for (int j = 0; j < board[i].Length; j++)
					{
						if (IsAliveCell(i, j))
						{
							const int PopulationMin = 2;
							const int PopulationMax = 3;
							int neighbourCount = GetNeighbourCount(i, j);

							if (neighbourCount < PopulationMin || neighbourCount > PopulationMax)
							{
								deadCells.Add(new Cell(i, j));
							}
						}
					}
				}

				for (int i = 0; i < board.Count; i++)
				{
					for (int j = 0; j < board[i].Length; j++)
					{
						if (!IsAliveCell(i, j))
						{
							const int RebirthCount = 3;
							int neighbourCount = GetNeighbourCount(i, j);

							if (neighbourCount == RebirthCount)
							{
								newCells.Add(new Cell(i, j));
							}
						}
					}
				}

				foreach (Cell cell in newCells)
				{
					board[cell.x][cell.y] = (char)CellStates.Alive;
				}

				foreach (Cell cell in deadCells)
				{
					board[cell.x][cell.y] = (char)CellStates.Dead;
				}

				cells = CreateCellsFromBoard();
			}
			else
			{
				board.Add(new StringBuilder((char)CellStates.Dead, 1));
			}

			return board;
		}

		private void CellsToBoard()
		{
			board = new List<StringBuilder>();

			for (int i = 0; i < Cell.maxX - Cell.minX + 1; i++)
			{
				string row = new string((char)CellStates.Dead, Cell.maxY - Cell.minY + 1);
				board.Add(new StringBuilder(row));
			}

			foreach (Cell cell in cells)
			{
				board[cell.x - Cell.minX][cell.y - Cell.minY] = (char)CellStates.Alive;
			}
		}

		private List<Cell> CreateCellsFromBoard()
		{
			List<Cell> newCells = new List<Cell>();

			for (int i = 0; i < board.Count; i++)
			{
				for (int j = 0; j < board[i].Length; j++)
				{
					if (IsAliveCell(i, j))
					{
						newCells.Add(new Cell(i, j));
					}
				}
			}

			return newCells;
		}

		private void UpdateCellMinAndMax()
		{
			Cell.minX = int.MaxValue;
			Cell.maxX = int.MinValue;
			Cell.minY = int.MaxValue;
			Cell.maxY = int.MinValue;

			foreach (Cell cell in cells)
			{
				Cell.minX = Math.Min(cell.x, Cell.minX);
				Cell.maxX = Math.Max(cell.x, Cell.maxX);
				Cell.minY = Math.Min(cell.y, Cell.minY);
				Cell.maxY = Math.Max(cell.y, Cell.maxY);
			}

			Cell.minX--;
			Cell.maxX++;
			Cell.minY--;
			Cell.maxY++;
		}

		private int GetNeighbourCount( int x, int y)
		{
			int count = 0;

			if (x > 0)
			{
				if (y > 0)
				{
					if (IsAliveCell(x - 1, y - 1))
					{
						count++;
					}
				}

				if (IsAliveCell(x - 1, y))
				{
					count++;
				}

				if (y < board[x].Length - 1)
				{
					if (IsAliveCell(x - 1, y + 1))
					{
						count++;
					}
				}
			}

			if (x < board.Count - 1)
			{
				if (y > 0)
				{
					if (IsAliveCell(x + 1, y - 1))
					{
						count++;
					}
				}

				if (IsAliveCell(x + 1, y))
				{
					count++;
				}

				if (y < board[x].Length - 1)
				{
					if (IsAliveCell(x + 1, y + 1))
					{
						count++;
					}
				}
			}

			if (y > 0)
			{
				if (IsAliveCell(x, y - 1))
				{
					count++;
				}
			}

			if (y < board[x].Length - 1)
			{
				if (IsAliveCell(x, y + 1))
				{
					count++;
				}
			}

			return count;
		}

		private void PrintBoard(int iteration = 0)
		{
			Console.Clear();
			Console.WriteLine($"Iteration {iteration}");

			foreach (StringBuilder row in board)
			{
				Console.WriteLine(row);
			}
		}

		private bool IsAliveCell(int x, int y) => board[x][y] == (char)CellStates.Alive;
	}

	internal class Day39
	{
		private static int Main(string[] args)
		{
			List<Cell> cells = GetCellsFromFile("Board.txt");
			GameOfLife gameOfLife = new GameOfLife(cells);

			gameOfLife.Play();

			Console.ReadLine();

			return 0;
		}

		private static List<Cell> GetCellsFromFile(string filename)
		{
			string[] lines = File.ReadAllLines(filename);
			List<Cell> cells = new List<Cell>();

			for (int i = 0; i < lines.Length; i++)
			{
				for (int j = 0; j < lines[i].Length; j++)
				{
					if (lines[i][j] == (char)CellStates.Alive)
					{
						cells.Add(new Cell(i, j));
					}
				}
			}

			return cells;
		}
	}
}

File: Days 031 - 040/Day 39/ConwaysGameOfLife.cpp
#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

enum class CellStates : char
{
	Alive = '*',
	Dead = '.'
};

struct Cell
{
	static int minX;
	static int maxX;
	static int minY;
	static int maxY;

	int x;
	int y;

	explicit Cell(int x, int y) noexcept
		: x(x), y(y)
	{ }

	friend bool operator <(const Cell& lhs, const Cell& rhs) noexcept;
	friend bool operator ==(const Cell& lhs, const Cell& rhs) noexcept;
};

int Cell::minX = std::numeric_limits<int>::max();
int Cell::maxX = std::numeric_limits<int>::min();
int Cell::minY = std::numeric_limits<int>::max();
int Cell::maxY = std::numeric_limits<int>::min();

bool operator <(const Cell& lhs, const Cell& rhs) noexcept
{
	return lhs.x != rhs.x ? lhs.x < rhs.x : lhs.y < rhs.y;
}

bool operator ==(const Cell& lhs, const Cell& rhs) noexcept
{
	return (lhs.x == rhs.x) && (lhs.y == rhs.y);
}

void PrintBoard(const std::vector<std::string>& board, unsigned int iteration = 0)
{
	static const std::string ScreenClearer(100, '\n');

	std::cout << ScreenClearer;
	std::cout << "Iteration " << iteration << "\n";

	for (const auto& row : board)
	{
		std::cout << row << "\n";
	}
}

unsigned int GetNeighbourCount(const std::vector<std::string>& board, unsigned int x, unsigned int y) noexcept
{
	unsigned int count = 0;

	for (unsigned int i = x - 1; i <= static_cast<int>(x + 1); i++)
	{
		for (unsigned int j = y - 1; j <= static_cast<int>(y + 1); j++)
		{
			if (i >= 0 && i < static_cast<int>(board.size()) &&
				j >= 0 && j < static_cast<int>(board[i].length()) &&
				!(i == x && j == y) && board[i][j] == static_cast<char>(CellStates::Alive))
			{
				count++;
			}
		}
	}

	return count;
}

void UpdateCellMinAndMax(const std::vector<Cell>& cells) noexcept
{
	Cell::minX = std::numeric_limits<int>::max();
	Cell::maxX = std::numeric_limits<int>::min();
	Cell::minY = std::numeric_limits<int>::max();
	Cell::maxY = std::numeric_limits<int>::min();

	for (const auto& cell : cells)
	{
		Cell::minX = std::min(cell.x, Cell::minX);
		Cell::maxX = std::max(cell.x, Cell::maxX);
		Cell::minY = std::min(cell.y, Cell::minY);
		Cell::maxY = std::max(cell.y, Cell::maxY);
	}

	Cell::minX--;
	Cell::maxX++;
	Cell::minY--;
	Cell::maxY++;
}

std::vector<std::string> CellsToBoard(const std::vector<Cell>& cells) noexcept
{
	std::vector<std::string> board;

	for (int i = 0; i < (Cell::maxX - Cell::minX + 1); i++)
	{
		board.emplace_back(Cell::maxY - Cell::minY + 1, static_cast<char>(CellStates::Dead));
	}

	for (const auto& cell : cells)
	{
		board[cell.x - Cell::minX][cell.y - Cell::minY] = static_cast<char>(CellStates::Alive);
	}

	return board;
}

std::vector<Cell> CreateCellsFromBoard(const std::vector<std::string>& board) noexcept
{
	std::vector<Cell> cells;

	for (unsigned int i = 0; i < board.size(); i++)
	{
		for (unsigned int j = 0; j < board[i].length(); j++)
		{
			if (board[i][j] == static_cast<char>(CellStates::Alive))
			{
				cells.emplace_back(i, j);
			}
		}
	}

	return cells;
}

std::vector<std::string> IterateCells(std::vector<Cell>& cells)
{
	UpdateCellMinAndMax(cells);

	std::vector<std::string> board;

	std::vector<Cell> newCells;
	std::vector<Cell> deadCells;

	if (!cells.empty())
	{
		board = CellsToBoard(cells);

		for (unsigned int i = 1; i < board.size() - 1; i++)
		{
			for (unsigned int j = 1; j < board[i].length() - 1; j++)
			{
				if (board[i][j] == static_cast<char>(CellStates::Alive))
				{
					unsigned int neighbourCount = GetNeighbourCount(board, i, j);

					if (constexpr unsigned int POPULATION_MIN = 2, POPULATION_MAX = 3; (neighbourCount < POPULATION_MIN) || (neighbourCount > POPULATION_MAX))
					{
						deadCells.emplace_back(i, j);
					}
				}
			}
		}

		for (unsigned int i = 0; i < board.size(); i++)
		{
			for (unsigned int j = 0; j < board[i].length(); j++)
			{
				if (board[i][j] == static_cast<char>(CellStates::Dead))
				{
					unsigned int neighbourCount = GetNeighbourCount(board, i, j);

					if (constexpr unsigned int REBIRTH_COUNT = 3; neighbourCount == REBIRTH_COUNT)
					{
						newCells.emplace_back(i, j);
					}
				}
			}
		}

		for (const auto& cell : newCells)
		{
			board[cell.x][cell.y] = static_cast<char>(CellStates::Alive);
		}

		for (const auto& cell : deadCells)
		{
			board[cell.x][cell.y] = static_cast<char>(CellStates::Dead);
		}

		cells = CreateCellsFromBoard(board);
	}
	else
	{
		board.emplace_back(1, static_cast<char>(CellStates::Dead));
	}

	return board;
}

void GameOfLife(const std::vector<Cell>& cells, unsigned int iterations = 100, unsigned int millisecondDelta = 500)
{
	UpdateCellMinAndMax(cells);
	std::vector<std::string> board = CellsToBoard(cells);

	PrintBoard(board);
	std::this_thread::sleep_for(std::chrono::milliseconds(millisecondDelta));

	std::vector<Cell> changingCells = cells;

	for (unsigned int i = 0; i < iterations; i++)
	{
		board = IterateCells(changingCells);
		PrintBoard(board, i + 1);

		std::this_thread::sleep_for(std::chrono::milliseconds(millisecondDelta));
	}
}

std::vector<Cell> GetCellsFromFile(const std::string& filename)
{
	std::ifstream cellFile(filename);

	if (!cellFile.is_open())
	{
		throw std::invalid_argument("File " + filename + " could not be found/opened.");
	}

	std::vector<Cell> cells;

	std::string currentLine;
	unsigned int lineCounter = 0;

	while (std::getline(cellFile, currentLine))
	{
		for (unsigned int i = 0; i < currentLine.length(); i++)
		{
			if (currentLine[i] == static_cast<char>(CellStates::Alive))
			{
				cells.emplace_back(lineCounter, i);
			}
		}

		lineCounter++;
	}

	return cells;
}

int main(int argc, const char* argv[])
{
	auto cells = GetCellsFromFile("Board.txt");
	GameOfLife(cells);

	std::cin.get();

	return 0;
}
